pkg/commands/waiter.go

```diff
@@ -0,0 +1,183 @@
+package commands
+
+import (
+	"context"
+	"fmt"
+	"github.com/pivotal/kpack/pkg/apis/build/v1alpha1"
+	corev1alpha1 "github.com/pivotal/kpack/pkg/apis/core/v1alpha1"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	watchTools "k8s.io/client-go/tools/watch"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/pkg/kmeta"
+)
+
+type ResourceWaiter interface {
+	Wait(runtime.Object) error
+	BuilderWait(runtime.Object, int64, int64) error
+}
+
+func NewResourceWaiter(dc dynamic.Interface) ResourceWaiter {
+	return NewWaiter(dc)
+}
+
+type Waiter struct {
+	dynamicClient dynamic.Interface
+}
+
+func NewWaiter(dc dynamic.Interface) *Waiter {
+	return &Waiter{dynamicClient: dc}
+}
+
+func (i *Waiter) Wait(ob runtime.Object) error {
+	m, ok := ob.(kmeta.OwnerRefable)
+	if !ok {
+		return errors.New("unexpected type")
+	}
+	return i.wait(ob, hasResolved(m.GetObjectMeta().GetGeneration()))
+}
+
+// Used to ensure the builder has resolved to the expected stack and store (importing)
+func (i *Waiter) BuilderWait(ob runtime.Object, expectedStoreGeneration, expectedStackGeneration int64) error {
+	m, ok := ob.(kmeta.OwnerRefable)
+	if !ok {
+		return errors.New("unexpected type")
+	}
+	return i.wait(ob, builderHasResolved(m.GetObjectMeta().GetGeneration(), expectedStoreGeneration, expectedStackGeneration))
+}
+
+func (i *Waiter) wait(ob runtime.Object, cf watchTools.ConditionFunc) error {
+	e := &watch.Event{Object: ob}
+	done, err := cf(*e)
+	if err != nil {
+		return err
+	}
+	if !done {
+		m, ok := ob.(kmeta.OwnerRefable)
+		if !ok {
+			return errors.New("unexpected type")
+		}
+		name := m.GetObjectMeta().GetName()
+		gvr, _ := meta.UnsafeGuessKindToResource(m.GetGroupVersionKind())
```

Comment:
assume this is not a problem for our resources.. (interesting function name 😁 )
```
// UnsafeGuessKindToResource converts Kind to a resource name.
// Broken. This method only "sort of" works when used outside of this package.  It assumes that Kinds and Resources match
// and they aren't guaranteed to do so.
func UnsafeGuessKindToResource(kind schema.GroupVersionKind) ( /*plural*/ schema.GroupVersionResource /*singular*/, schema.GroupVersionResource) {
...
```

Reply:
good question, at least we know we'll be using our resources which do have matching kinds. I've looked around and this seems to be the way to get the GroupVersionResource